File: amqnullmessages/src/main/scala/codemettle/TopicListeners.scala
package codemettle

import akka.actor.{ActorRef, ActorSystem, Props}
import codemettle.TopicListeners.TopicListener
import com.codemettle.reactivemq.TopicConsumer
import com.codemettle.reactivemq.model.{AMQMessage, Topic}
import com.typesafe.scalalogging.Logger
import org.slf4j.LoggerFactory

object TopicListeners {
  class TopicListener(doLogging: Boolean, val consumeFrom: Topic) extends TopicConsumer {
    private lazy val logger = Logger(LoggerFactory.getLogger(consumeFrom.toString))

    override def connection: ActorRef = esbConnection

    override def receive: Receive = {
      case m@AMQMessage(null, _, _) if doLogging =>
        logger.error(s"Got NULL message body: $m")

      case m if doLogging => logger.trace(s"got message $m")
      
      case _ => // ignore
    }
  }
}

class TopicListeners(val system: ActorSystem, doLogging: Boolean) extends HasSystem {
  system.actorOf(Props(new TopicListener(doLogging, Topic(Topic1))))
  system.actorOf(Props(new TopicListener(doLogging, Topic(Topic2))))
  system.actorOf(Props(new TopicListener(doLogging, Topic(Topic3))))
}

File: amqnullmessages/src/main/java/com/codemettle/TopicListeners.java
package com.codemettle;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import javax.jms.JMSException;
import javax.jms.MessageConsumer;
import javax.jms.TextMessage;
import java.io.Closeable;
import java.io.IOException;

import static com.codemettle.Util.*;

public class TopicListeners implements Closeable {
    public static class TopicListener implements Closeable {
        private final Logger m_logger = LoggerFactory.getLogger(getClass());
        private final MessageConsumer m_cons;

        TopicListener(final boolean aDoLogging, final String aConsumeFrom,
                      final CloseableSession aSession) throws JMSException {
            m_cons = aSession.getSession().createConsumer(aSession.getSession().createTopic(aConsumeFrom));
            m_cons.setMessageListener(message -> {
                if (aDoLogging) {
                    if (message instanceof TextMessage) {
                        try {
                            if (((TextMessage) message).getText() == null) {
                                m_logger.error("Got NULL message body: " + message);
                            } else {
                                m_logger.trace("got message " + message);
                            }
                        } catch (final JMSException e) {
                            m_logger.error("Error", e);
                        }
                    } else {
                        m_logger.warn("Got unexpected message: " + message);
                    }
                }
            });
        }

        @SuppressWarnings("RedundantThrows")
        @Override
        public void close() throws IOException {
            try {
                m_cons.close();
            } catch (JMSException e) {
                // ignore
            }
        }
    }

    private final TopicListener m_top1, m_top2, m_top3;

    TopicListeners(final CloseableSession aSession, final boolean aDoLogging) throws JMSException {
        m_top1 = new TopicListener(aDoLogging, TOPIC_1, aSession);
        m_top2 = new TopicListener(aDoLogging, TOPIC_2, aSession);
        m_top3 = new TopicListener(aDoLogging, TOPIC_3, aSession);
    }

    @Override
    public void close() throws IOException {
        m_top1.close();
        m_top2.close();
        m_top3.close();
    }
}
